tripleawarclub.org/public_html/dice/dice.class.php

```diff
@@ -0,0 +1,287 @@
+		$result = $this->dbconn->query($sql) or exit("fatal error: data connection lost @checkIfMailsAreRegistered!");
+		$registered_mails = mysqli_fetch_array($result);
+		$num_emails = $result->num_rows;
+		if($num_emails == count($emails)){
+			return true;	// all emails are registered
+		}
+		if($registered_mails == false){
+			throw new exception("fatal error: none of the emails is registered. Please register emails at ".$this->domain."/register.php !");
+		}
+
+		foreach($emails as $email) {
+			if(! in_array($email, $registered_mails)){
+				throw new exception("fatal error: email $email is not registered. Please register email at ".$this->domain."/register.php !");
+			}
+		}
+		throw new exception("fatal error: unknown error with email adresses!");
+	}
+
+	/**
+	 * Checks if a specific email is already registered
+	 * @return bool Is the email registered
+	 */
+	function checkIfMailIsRegistered($email) {
+		$this->connectDatabase();
+
+		$sql = "SELECT registered_email FROM dice_emails WHERE registered_email = '$email'";
+		$result = $this->dbconn->query($sql) or exit("fatal error: data connection error " . $this->dbconn->error . "!");
+		$num_emails = $result->num_rows;
+
+		return ($num_emails == 1);
+	}
+
+	/**
+	 * runs any SQL query on the database
+	 * @param string $sql
+	 * @return mixed mysqlressource
+	 */
+	function runQuery($sql) {
+		$this->connectDatabase();
+
+		$result = $this->dbconn->query($sql) or exit("fatal error: " . $this->dbconn->error . "!");
+		return $result;
+	}
+
+	function runStatement($sth){
+		$this->connectDatabase();
+		$result = $sth->execute();
+		return $result;
+	}
+    function disconnectDatabase() {
+    	mysqli_close($this->dbconn);
+    }
+
+	////////////////////////////////
+	//        Encryption          //
+	////////////////////////////////
+	/**
+	 * returns the date and key
+	 * if no date is specified the latest key in key.dat will be returned
+	 */
+	function getEncryptionKey($date = null) {
+		// get old key
+		if($date) {
+			$dir = dirname(__FILE__);
+			$keyfile = fopen("$dir/keys/$date.dat", "r");
+		} else {
+			// get current key
+			if (!file_exists("key.dat")) {
+				$this->setEncryptionKey();
+			}
+			$keyfile = fopen("key.dat", "r");
+		}
+
+		if ($keyfile) {
+			$data = fread($keyfile, 8192);
+
+			$this->enc = unserialize($data);
+			fclose($keyfile);
+
+			return $this->enc;
+		} else {
+			exit("fatal error: Wrong date!");
+		}
+	}
+
+	function setEncryptionKey() {
+		$shouldWriteKey = true;
+		if (file_exists("key.dat")) {
+			$old = $this->getEncryptionKey();
+			$shouldWriteKey = copy("key.dat", "./keys/" . $old['date'] . ".dat");
+		}
+
+		if ($shouldWriteKey) {
+			$file = fopen("key.dat", "w");
+			$enc['key'] = $this->keygen();
+			$enc['date'] = $this->getDate();
+			$output = serialize($enc);
+			fputs($file, $output);
+			fclose($file);
+			chmod("key.dat", 0600);
+		}
+	}
+
+    function keygen() {
```

Comment:
I believe there's a more efficient way of generating a random password of length 24 matching `[A-Za-z0-9]{24}`